### ISSUE TITLE
The Python Training API should not alter "params" argument in place

### ISSUE THREAD

**Issue by Mithrillion** [Microsoft/LightGBM#1706]
It appears that the training API's ` train()` method will attempt to modify the `params` argument in place with `pop()`, new key assignment, etc. when trying to use the categorical columns from a Pandas dataframe. This is likely to cause unexpected behaviour when the `params` values are reused. For instance, I have the following params dict:
```
reg_params = {
        "objective": "regression",
        "metric": "rmse",
        "max_depth": 10,
        "min_child_samples": 20,
        "lambda_l1": 1,
        "lambda_l2": 1,
        "num_leaves": 150,
        "num_iterations": 20000,
        "learning_rate": 0.01,
        "subsample": 0.8,
        "colsample_bytree": 0.8,
        "verbosity": -1
    }
```
And I used the dict in a KFold setup in which I used the same parameters in each fold. After one fold, the dict becomes:
`{'objective': 'regression', 'metric': 'rmse', 'max_depth': 10, 'min_child_samples': 20, 'lambda_l1': 1, 'lambda_l2': 1, 'num_leaves': 150, 'learning_rate': 0.01, 'subsample': 0.8, 'colsample_bytree': 0.8, 'verbosity': -1, 'categorical_column': [0, 2, 3, 5, 6, 7, 8, 9, 14, 17, 18, 20, 21, 22, 27, 28, 29]}`
Notice that the `num_iterations` value is lost despite not altering the dict manually anywhere.

The expected behaviour should be that the API reads the params dict and makes a run-specific copy of it instead of altering its value in place. Something like
```
def train(og_params, *):
    params = og_params.copy()
```
should do the trick.

**Comment by StrikerRUS** [comment]
@Mithrillion You are absolutely right! To be honest, I've noticed that there are many places where some arguments are modified in LightGBM's functions, which causes some side effects. I think this is unacceptable and should be fixed. I'll try to find all such places and fix them.
